request.go

```diff
@@ -42,7 +43,68 @@ func AllowsBody(r *http.Request) bool {
 
 // HasBody returns true if this method needs a content-type
 func HasBody(r *http.Request) bool {
-	return len(r.TransferEncoding) > 0 || r.ContentLength > 0
+	// happy case: we have a content length set
+	if r.ContentLength > 0 {
+		return true
+	}
+
+	if r.Header.Get(http.CanonicalHeaderKey("content-length")) != "" {
+		// in this case, no Transfer-Encoding should be present
+		// we have a header set but it was explicitly set to 0, so we assume no body
+		return false
+	}
+
+	rdr := newPeekingReader(r.Body)
```

Comment:
my concern below is invalid if `r.Body` represents an entity-body after all transfer-encodings have been applied. but if it does not, then I think the `peek(1)` will observe ascii "0" (end of chunk marker) and mistakenly assume there is a body.

Reply:
the go stdlib takes care of this part